lib/gc/ExecutionEngine/CPURuntime/MemoryPool.cpp

```diff
@@ -0,0 +1,277 @@
+    return divideAndCeil(sz, getOsPageSize()) * getOsPageSize();
+  } else {
+    return blockSize;
+  }
+}
+
+void *FILOMemoryPool::alloc(size_t sz) {
+  if (unlikely(!buffers)) {
+    buffers = MemoryBlock::make(getBlockSize(sz), nullptr, nullptr);
+    current = buffers;
+  }
+  do {
+    intptr_t newptr = current->callAllocPtr();
+    size_t newallocated = newptr + sz - reinterpret_cast<intptr_t>(current);
+    if (likely(newallocated <= current->size)) {
+      // if the current block is not full
+      size_t alloc_size = newallocated - current->allocated;
+      current->allocated = newallocated;
+      MemoryChunk::init(newptr, alloc_size, poolType);
+      return reinterpret_cast<void *>(newptr);
+    }
+    // if the block is full, check the next block
+    // if there is no next block left, allocate a new one
+    if (!current->next) {
+      current->next = MemoryBlock::make(getBlockSize(sz), current, nullptr);
+    }
+    current = current->next;
+  } while (true);
+}
+
+void FILOMemoryPool::dealloc(void *ptr) {
+  auto intptr = reinterpret_cast<intptr_t>(ptr);
+  auto intcur = reinterpret_cast<intptr_t>(current);
+  // Optional: check if the pointer is valid in the current block
+
+  assert(intptr > intcur &&
+         intptr - intcur < static_cast<ptrdiff_t>(current->size));
+  auto chunk = reinterpret_cast<MemoryChunk *>(intptr - sizeof(MemoryChunk));
+  // Optional: check if the stack is ok
+  assert(chunk->canary == MemoryChunk::magicCheckNum &&
+         "Corrupt stack detected");
+  assert(current->allocated > chunk->size);
+  current->allocated -= chunk->size;
+
+  // skip the empty blocks
+  while (unlikely(current->allocated == sizeof(MemoryBlock))) {
+    if (current->prev) {
+      current = current->prev;
+    } else {
+      break;
+    }
+  }
+}
+
+void FILOMemoryPool::release() {
+  freeMemoryBlockList(buffers);
+  buffers = nullptr;
+  current = nullptr;
+}
+
+void FILOMemoryPool::clear() {
+  for (auto cur = current; cur; cur = cur->prev) {
+    cur->allocated = sizeof(MemoryBlock);
+  }
+  current = buffers;
+}
+
+FILOMemoryPool::~FILOMemoryPool() { release(); }
+
+} // namespace
+
+static thread_local FILOMemoryPool mainMemoryPool_{mainChunkSize,
+                                                   MemoryPoolType::MAIN};
+// if the current thread is a worker thread, use this pool
+static thread_local FILOMemoryPool threadMemoryPool_{threadlocalChunkSize,
+                                                     MemoryPoolType::THREAD};
+
+extern "C" void *gcAlignedMalloc(size_t sz, MemoryPoolType type) noexcept {
+  if (likely(type == MemoryPoolType::MAIN)) {
+    return mainMemoryPool_.alloc(sz);
+  } else {
+    return threadMemoryPool_.alloc(sz);
+  }
+}
+
+extern "C" void gcAlignedFree(void *p) noexcept {
+  auto chunk = reinterpret_cast<MemoryChunk *>(reinterpret_cast<intptr_t>(p) -
+                                               sizeof(MemoryChunk));
+  if (likely(chunk->poolType == MemoryPoolType::MAIN)) {
```

Comment:
not sure if it is a good design to dynamically decide the pool kind. 
1. This information is known at compile time
2. poolType takes memory for every allocation